[Feat] Button 구현

kamja0510 opened the pull request:
## Related issue 🛠
- closed #25 

## Work Description ✏️
- [x] TextButton
- [x] GuidelineButton
- [x] SmallGrayButton
- [x] StreamingButton
- [x] LargePinkButton
- [x] LargeGrayButton
- [x] KakaoLoginButton
- [x] BookmarkButton
- [x] LIkeButton
- [x] Switch
- [x] CircleButton

## Screenshot 📸
<table width="100%">
    <tr>
      <td width="40%">설명</td>
      <td width="60%">스크린샷</td>
    </tr>
    <tr>
      <td width="40%">
        CircleButton
      </td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/bcf099ae-35d8-4252-9595-7a211a81dcf8" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">Switch</td>
      <td width="60%">
        <video src="https://github.com/user-attachments/assets/a6e79af8-693f-44db-9652-99fc73fbf807">
      </td>
    </tr>
    <tr>
      <td width="40%">LIkeButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/7b17e9fb-6dcf-41ad-90b1-0d21a7bcb707" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">BookmarkButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/c2e544b0-9dc4-4ee5-81c7-ec4d8efe266b" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">KakaoLoginButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/94e3491f-8f55-488f-a56b-81d1d0b25911" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">LargeGrayButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/dcf88268-9af7-4376-8454-b65fa1ef8605" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">LargePinkButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/c9f0a450-05de-41a8-9a8b-0b288f1246e1" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">StreamingButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/b38d8ace-c925-4a76-add3-9db8d3384164" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">SmallGrayButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/ced3a6b6-5e15-4ac6-b93e-287351559baf" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">GuidelineButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/57148829-715a-4e5a-9505-e9a760b073ba" style="height:auto;">
      </td>
    </tr>
    <tr>
      <td width="40%">TextButton</td>
      <td width="60%">
        <img src="https://github.com/user-attachments/assets/935c6b63-6f37-49ce-99c5-e59ad6b0f8c5" style="height:auto;">
      </td>
    </tr>
</table>

## Uncompleted Tasks 😅

## To Reviewers 📢

kamja0510 committed:
[feat] #25 DPlayTextButton 구현

kamja0510 committed:
[feat] #25 DPlayCircleButton 구현

kamja0510 committed:
[feat] #25 DPlayBookmarkButton 구현

kamja0510 committed:
[feat] #25 DPlaySwitch 구현

kamja0510 committed:
[feat] #25 DPlayButtonSlot 구현

kamja0510 committed:
[feat] #25 DPlayStreamingButton 구현

kamja0510 committed:
[feat] #25 버튼에 onClick 붙임

kamja0510 committed:
[feat] #25 DplayLikeButton 구현

kamja0510 committed:
[refactor] #25 DPlayButtonSlot.kt 구조변경

kamja0510 committed:
[feat] #25 DPlayGuidelineButton 구현

kamja0510 committed:
[feat] #25 DPlaySmallGrayButton 구현

kamja0510 committed:
[feat] #25 DPlayLargePinkButton 구현

kamja0510 committed:
[feat] #25 DPlayLargeGrayButton 구현

kamja0510 committed:
[feat] #25 DPlayKakaoLoginButton 구현

kamja0510 committed:
[refactor] #25 DPlayButtonSlot 매개변수 위치 수정

kamja0510 committed:
[refactor] #25 DPlayBookmarkButton.kt DplayButtonSlot 사용하는 것으로 변경

kamja0510 committed:
[refactor] #25 문자열 추출 수정

kamja0510 committed:
[feat] #25 circle button 각각의 사용처 구현

kamja0510 committed:
[chore] #25 들여쓰기 지우기

kamja0510 committed:
[chore] #25 DPlayLikeButtonPreview 수정

kamja0510 committed:
[chore] #25 불필요한 import문 제거

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/DPlayBookmarkButton.kt:
```diff
@@ -0,0 +1,61 @@
+    val iconRes = if (isMarked) R.drawable.ic_bookmark_filled_24 else R.drawable.ic_bookmark_unfilled_24
+    val iconContentDescription = stringResource(
+        if (isMarked) R.string.bookmark_button_filled_bookmark_icon_description else R.string.bookmark_button_unfilled_bookmark_icon_description
+    )
```
```suggestion
    val (iconRes, iconContentDescription) = if (isMarked) {
    R.drawable.ic_bookmark_filled_24 to stringResource(R.string.bookmark_button_filled_bookmark_icon_description)
} else {
    R.drawable.ic_bookmark_unfilled_24 to stringResource(R.string.bookmark_button_unfilled_bookmark_icon_description)
}
```
구조분해 선언 suggestion 드립니다

kamja0510 replied:
오호 보통 이렇게 하는게 추천되는 패턴인가요?
뭔가 굳이 Pair객체로 붙인 다음에 분해 하는게 살짝 이상해보여서요.
그럼 3개면 Triple로 붙였다가 다시 분해하고(밑에 코드리뷰에 달려있네요.)
4개 이상이면 list로 만들었다가 다시 분해하나요??

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/button/DPlayBookmarkButton.kt:
```diff
@@ -0,0 +1,61 @@
+        Icon(
+            imageVector = ImageVector.vectorResource(iconRes),
+            contentDescription = iconContentDescription,
+            tint = DPlayTheme.colors.dplayWhite
```
```suggestion
            tint = Color.Unspecified
```

이미 dplayWhite 색상으로 추출되어있는걸로 아는데 
벡터의 path fill color를 다시한번 흰색으로 덮게 되어서 만약 추출된 아이콘과 다른색을 의도적으로 입히는 경우가 아니라면
벡터가 가진 원본 색상 그대로 렌더링하는것이 더 좋다고 생각합니다!

kamja0510 replied:
저도 처음에 명시성 vs 성능 고민하다가 성능이 미세하다고 하여 명시적인게 좋지 않을까 했다가 뒤에부턴 귀찮아서 안했는데 그냥 Unspecified로 고정하도록 하겠습니다.

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/DPlayKakaoLoginButton.kt:
```diff
@@ -0,0 +1,78 @@
+    modifier : Modifier = Modifier,
+    onClick: () -> Unit = {}
+){
+    DPlayButtonSlot(
+        modifier = modifier,
+        onClick = onClick,
+        paddingValues = PaddingValues(20.dp),
+        containerColor = DPlayTheme.colors.kakaoYellow,
+        borderColor = DPlayTheme.colors.kakaoYellow,
+    ) {
+        Row(
```
DPlayButtonSlot 구현쪽을 보니 Row 내부에 content()를 넣는 구조인 것 같은데 이렇게 사용하는 경우에는 Row()가 중첩되지는 않나요?

kamja0510 replied:
처음에 슬롯 안에 trailing content, leading content도 있었어서 Slot 내부에 Row를 넣어놨었는데 매의눈 ㄷㄷ

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/DPlayButtonSlot.kt:
```diff
@@ -0,0 +1,55 @@
+import com.example.designsystem.util.noRippleClickable
+
+@Composable
+fun DPlayButtonSlot(
```
피그마에 맞게 ActionButton 네이밍은 어떻게 생각하시나요?

kamja0510 replied:
이게 ActionButton 카테고리 말고 ToggleButton에도 쓰여서요 그냥 기본 slot이라고 생각하시면 됩니다

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/DPlayLargeGrayButton.kt:
```diff
@@ -0,0 +1,73 @@
+    onClick: () -> Unit = {}
+){
+    val containerColor = if(enabled) DPlayTheme.colors.gray600 else DPlayTheme.colors.gray200
+    val textColor = if(enabled) DPlayTheme.colors.dplayWhite else DPlayTheme.colors.gray400
```
여기도 구조분해 선언 권장!

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/button/DPlayLikeButton.kt:
```diff
@@ -0,0 +1,94 @@
+fun DPlayLikeButton(
+    modifier: Modifier = Modifier,
+    likeCount: Int = 0,
+    isLiked: Boolean = false,
```
중요한 건 아니지만 optional보다는 필수인자여야 하지않나? 라는 의견드려봅니당

kamja0510 replied:
중요하네요~~
근데 생각해보니 button 컴포넌트면 onClick도 필수인자겠네요

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/DPlayLikeButton.kt:
```diff
@@ -0,0 +1,94 @@
+){
+    val containerColor = if(isLiked) DPlayTheme.colors.dplayPink100 else DPlayTheme.colors.dplayWhite
+    val iconRes = if(isLiked) R.drawable.ic_heart_pink_filled_24 else R.drawable.ic_heart_pink_unfilled_24
+    val stringRes = if(isLiked) R.string.like_button_filled_heart_description else R.string.like_button_unfilled_heart_description
```
여기도 Triple을 이용한 ㄱㅈㅂㅎ

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/DPlayStreamingButton.kt:
```diff
@@ -0,0 +1,89 @@
+            ){
+                Icon(
+                    imageVector = ImageVector.vectorResource(R.drawable.ic_stream_white_32),
+                    contentDescription = stringResource(R.string.streaming_button_icon_description),
```
ㄷㄷ슨

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/button/DPlaySwitch.kt:
```diff
@@ -0,0 +1,113 @@
+import com.example.designsystem.util.noRippleClickable
+
+@Composable
+fun DPlaySwitch(
```
토글?

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/button/DPlaySwitch.kt:
```diff
@@ -0,0 +1,113 @@
+@Composable
+fun DPlaySwitch(
+    modifier: Modifier = Modifier,
+    isChecked: Boolean,
```
```suggestion
    isChecked: Boolean,
    modifier: Modifier = Modifier
```

kamja0510 replied:
굳 ㅋㅋ

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/button/DPlaySwitch.kt:
```diff
@@ -0,0 +1,113 @@
+    modifier: Modifier = Modifier,
+    isChecked: Boolean,
+    sizes: SwitchSizes = DPlaySwitchDefaults.sizes(),
+    colors: SwitchColors = DPlaySwitchDefaults.colors(),
```
어 귀찮으실까봐 안달았었는데 위에 enabled 나 isChecked 로 관리하셧던 변수들 다 이렇게 하면 굉장히 좋을 것 같습니다.... ㅎㅎ

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/button/DPlayTextButton.kt:
```diff
@@ -0,0 +1,49 @@
+import com.example.designsystem.util.noRippleClickable
+
+@Composable
+fun DPlayTextButton(
```
```suggestion
fun DPlayUnderlineTextButton(
```

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/constant/DPlayToggleDefaults.kt:
```diff
@@ -0,0 +1,48 @@
+        padding = padding,
+        cornerRadius = cornerRadius,
+        thumbSize = thumbSize
+    )
```
위에 말씀드린 것 처럼 버튼 전반적인 크기와 플래그에 따른 색상값들을 버튼의 타입에 따라 구분하여 정의해서 사용하면 가독성과 유지보수할 때 편할것 같습니다!

t1nm1ksun commented on core/designsystem/src/main/java/com/example/designsystem/component/button/type/CircleButtonType.kt:
```diff
@@ -0,0 +1,56 @@
+import com.dplay.designsystem.R
+import com.example.designsystem.theme.defaultDPlayColors
+
+sealed class CircleButtonType {
```
질문! 여긴 Immutable 어노테이션 불필요한가요?

kamja0510 replied:
inspection 찍어보니까 stable이긴한데 이런 타입 같은 것들에 명시적으로 붙여주는 거 좋은거 같습니다

t1nm1ksun reviewed (approved):
양이 엄청많네요 고생하셨습니다!
네이밍적인 부분은 최대한 피그마에 명시된 네이밍을 따르는게 소통하기 편할 것 같아요~

kamja0510 committed:
[merge] #25 branch 'develop' into feat/#25-button

# Conflicts:
#	core/designsystem/src/main/java/com/example/designsystem/util/ModifierExt.kt

kamja0510 committed:
[refactor] #25 컴포넌트 패키지로 옮기고 변수명 변경

kamja0510 committed:
[feat] #25 CircleButtonType에 immutable 어노테이션 붙이기

kamja0510 committed:
[refactor] #25 DPlayBookmarkButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlayCircleButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlayGuidelineButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlayKakaoLoginButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlayLargeGrayButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlayLargePinkButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlayLikeButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlaySmallGrayButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정

kamja0510 committed:
[refactor] #25 DPlayStreamingButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정
2. DPlayBaseIcon 쓰는 것으로 변경

kamja0510 committed:
[refactor] #25 DPlayToggle.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정

kamja0510 committed:
[refactor] #25 DPlayUnderlineTextButton.kt 코드리뷰 반영

1. onclick 필수 parameter로 설정

kamja0510 committed:
[refactor] #25 DPlayButtonSlot에 row없애기

kamja0510 committed:
[chore] #25 DPlaySwitchDefaults.kt -> DPlayToggleDefaults.kt 네이밍 변경

kamja0510 committed:
[Merge] #25 branch 'develop' into feat/#25-button

# Conflicts:
#	core/designsystem/src/main/java/com/example/designsystem/util/ModifierExt.kt

kamja0510 committed:
[feat] #25 GuidelineButton shape 수정

kamja0510 committed:
[feat] #25 BookmarkButton shape 수정

kamja0510 committed:
[feat] #25 KakaoLoginButton shape 수정

kamja0510 committed:
[feat] #25 slot shape RoundedCornerShape(12.dp)로 수정 후 12.dp가 아닌 것들 수정

kamja0510 committed:
[chore] #25 ktlintFormat
